respondo/jacobi_diis.py

```diff
@@ -0,0 +1,158 @@
+        print("Niter residual_norm", file=file)
+    elif identifier == "next_iter":
+        fmt = "{n_iter:3d}  {residual:12.5g}"
+        print(fmt.format(n_iter=state.n_iter,
+                         residual=np.max(state.residual_norm)), file=file)
+    elif identifier == "is_converged":
+        print("=== Converged ===", file=file)
+        print("    Number of matrix applies:   ", state.n_applies)
+
+
+def jacobi_diis(matrix, rhs, x0, Dinv, conv_tol=1e-9, max_iter=100,
+                callback=None, explicit_symmetrisation=IndexSymmetrisation,
+                max_subspace=8):
+    """An implementation of the Jacobi-DIIS algorithm.
+    It solves `matrix @ x = rhs` for `x` by minimising the residual
+    `matrix @ x - rhs` using a Jacobi update with DIIS acceleration.
+    Parameters
+    ----------
+    matrix
+        Matrix object. Should be an ADC matrix.
+    rhs
+        Right-hand side, source.
+    x0
+        Initial guess
+    Dinv
+        Inverse of the matrix diagonal D^{-1}
+    conv_tol : float
+        Convergence tolerance on the l2 norm of residuals to consider
+        them converged.
+    max_iter : int
+        Maximum number of iterations
+    callback
+        Callback to call after each iteration
+    explicit_symmetrisation
+        Explicit symmetrisation to perform during iteration to ensure
+        obtaining an eigenvector with matching symmetry criteria.
+    """
+    if callback is None:
+        def callback(state, identifier):
+            pass
+
+    if explicit_symmetrisation is not None and \
+            isinstance(explicit_symmetrisation, type):
+        explicit_symmetrisation = explicit_symmetrisation(matrix)
+
+    if x0 is None:
+        raise NotImplementedError("Random guess is not yet implemented.")
+    else:
+        x0 = copy(x0)
+
+    if Dinv is not None and isinstance(Dinv, type):
+        Dinv = Dinv(matrix)
+
+    def is_converged(state):
+        state.converged = state.residual_norm < conv_tol
+        return state.converged
+
+    state = State()
+    # Initialise iterates
+    state.solution = x0
+    if explicit_symmetrisation:
+        state.solution = explicit_symmetrisation.symmetrise(state.solution)
+    state.residual = evaluate(rhs - matrix @ state.solution)
+    state.residual_norm = np.sqrt(state.residual @ state.residual)
+
+    callback(state, "start")
+    diis = DIIS(max_subspace=max_subspace)
+    while state.n_iter < max_iter:
+        state.n_iter += 1
+
+        callback(state, "next_iter")
+        if is_converged(state):
+            state.converged = True
+            callback(state, "is_converged")
+            return state
+
+        if state.n_iter == max_iter:
+            raise la.LinAlgError("Maximum number of iterations (== "
+                                 + str(max_iter) + " reached in Jacobi/DIIS "
+                                 "procedure.")
+
+        x_new = evaluate(state.solution + Dinv @ state.residual)
```

Comment:
We were wondering if it could be done that way:
```
jacobi = evaluate(Dinv @ state.residual)
if explicit_symmetrisation:
    jacobi = explicit_symmetrisation.symmetrise(jacobi)
x_new = evaluate(state.solution + jacobi)
state.solution = diis.extrapolate(x_new, jacobi)
```
This way, in line 107, it would not be necessary to reverse the step from line 104, so to speak. But I don't know if it makes that much of a difference.